Sources/CartonHelpers/DiagnosticsParser.swift

```diff
@@ -0,0 +1,211 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+import TSCBasic
+import Splash
+
+fileprivate extension StringProtocol {
+  func matches(regex: NSRegularExpression) -> String.SubSequence? {
+    let str = String(self)
+    let range = NSRange(location: 0, length: utf16.count)
+    guard let match = regex.firstMatch(in: str, options: [], range: range),
+          let matchRange = Range(match.range, in: String(self)) else {
+      return nil
+    }
+    return str[matchRange.upperBound..<str.endIndex]
+  }
+}
+
+fileprivate extension String.StringInterpolation {
+  mutating func appendInterpolation<T>(_ value: T, color: String...) {
+    appendInterpolation("\(color.map { "\u{001B}\($0)" }.joined())\(value)\u{001B}[0m")
+  }
+}
+
+fileprivate extension TokenType {
+  var color: String {
+    // Reference on escape codes: https://en.wikipedia.org/wiki/ANSI_escape_code#Colors
+    switch self {
+    case .keyword: return "[35;1m" // magenta;bold
+    case .comment: return "[90m" // bright black
+    case .call, .dotAccess, .property, .type: return "[94m" // bright blue
+    case .number, .preprocessing: return "[33m" // yellow
+    case .string: return "[91;1m" // bright red;bold
+    default: return "[0m" // reset
+    }
+  }
+}
+
+fileprivate struct TerminalOutputFormat: OutputFormat {
+  func makeBuilder() -> TerminalOutputBuilder {
+    .init()
+  }
+  
+  struct TerminalOutputBuilder: OutputBuilder {
+    var output: String = ""
+    
+    mutating func addToken(_ token: String, ofType type: TokenType) {
+      output.append("\(token, color: type.color)")
+    }
+    
+    mutating func addPlainText(_ text: String) {
+      output.append(text)
+    }
+    
+    mutating func addWhitespace(_ whitespace: String) {
+      output.append(whitespace)
+    }
+    
+    mutating func build() -> String {
+      output
+    }
+  }
+}
+
+/// Parses and re-formats diagnostics output by the Swift compiler.
+///
+/// The compiler output often repeats iteself, and the diagnostics can sometimes be difficult to read.
+/// This reformats them to a more readable output.
+struct DiagnosticsParser {
+  enum Regex {
+    /// The output has moved to a new file
+    static let enterFile = try! NSRegularExpression(pattern: #"\[\d+\/\d+\] Compiling \w+ "#)
+    /// A message is beginning with the line # following the `:`
+    static let line = try! NSRegularExpression(pattern: #"(\/\w+)+\.\w+:"#)
+  }
+  
+  struct CustomDiagnostic {
+    let kind: Kind
+    let line: String.SubSequence
+    let char: String.SubSequence
+    let code: String
+    let message: String.SubSequence
+    
```

Comment:
<!--
  1 failure:  Lines should not ...
  0 warning: 
  
  
  DangerID: danger-id-Danger;
  File: Sources/CartonHelpers/DiagnosticsParser.swift;
  Line: 95;
-->
- :no_entry_sign: Lines should not have trailing whitespace. (`trailing_whitespace`)